File: src/org/usfirst/frc/team4099/robot/subsystems/Climber.java
package org.usfirst.frc.team4099.robot.subsystems;

import edu.wpi.first.wpilibj.Talon;
import org.usfirst.frc.team4099.lib.joystick.JoystickUtils;
import org.usfirst.frc.team4099.robot.loops.Loop;

/**
 * Created by plato2000 on 2/13/17.
 */
public class Climber implements Subsystem {

    public static Climber sClimber = new Climber();
    private Talon climberTalon;
    private double currentPower;


    private Climber() {
        this.climberTalon = new Talon(5);
    }

    public static Climber getInstance() {
        return sClimber;
    }

    @Override
    public void outputToSmartDashboard() {
//        SmartDashboard.putNumber("climberPower", );

    }

    @Override
    public synchronized void stop() {
        setClimberPower(0);
        climberTalon.set(0);
    }

    @Override
    public void zeroSensors() {}

    public synchronized void setClimberPower(double climberPower) {
        this.currentPower = Math.abs(JoystickUtils.deadband(climberPower, .2));
    }

    public Loop getLoop() {
        return mLoop;
    }

    private final Loop mLoop = new Loop() {
        @Override
        public void onStart() {
            setClimberPower(0);
        }

        @Override
        public void onLoop() {
            synchronized (Climber.this) {
                climberTalon.set(currentPower);
            }
        }

        @Override
        public void onStop() {
            stop();
        }
    };
}

File: src/main/kotlin/org/usfirst/frc/team4099/robot/subsystems/Climber.kt
package org.usfirst.frc.team4099.robot.subsystems

import com.ctre.phoenix.motorcontrol.ControlMode
import edu.wpi.first.wpilibj.Talon
import org.usfirst.frc.team4099.lib.util.CANMotorControllerFactory
import org.usfirst.frc.team4099.robot.Constants
import org.usfirst.frc.team4099.robot.loops.Loop


class Climber private constructor(): Subsystem {
    enum class ClimberState {
        CLIMBING, NOT_CLIMBING, UNCLIMBING
    }

    var climberState = ClimberState.NOT_CLIMBING

    val talon = Talon(Constants.Climber.CLIMBER_TALON_ID)

    override fun outputToSmartDashboard() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun stop() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun zeroSensors() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    val loop: Loop = object : Loop {
        override fun onStart() {
        }

        override fun onLoop() {
            synchronized(this@Climber) {
                when(climberState) {
                    ClimberState.CLIMBING -> talon.set(-1.0)
                    ClimberState.NOT_CLIMBING -> talon.set(0.0)
                    ClimberState.UNCLIMBING -> talon.set(1.0)
                }
            }
        }

        override fun onStop() {
        }
    }

    companion object {
        val instance = Climber()
    }

}